### PR TITLE
correct K2 KL implementation, and apply importance sampling to it

### PR DESCRIPTION
**Pull request by bbartoldson**
This corrects the Kimi K2 KL implementation. 

---

**GitHub Issue**: #1390 

---

> [!NOTE]
> Adjust KL regularizer in `compute_loss` to an importance-sampled K2 KL using a detached summed log-ratio scaled by per-token importance ratios.
> 
> - **RL Loss** (`src/prime_rl/trainer/rl/loss.py`):
>   - Update KL term in `compute_loss` to importance-sampled form: `kl_tau * (log_importance_ratio[kl_mask].sum().detach() * importance_ratio[kl_mask]).sum()` replacing simple sum of `log_importance_ratio`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 761aef48f3b4a7f8d92f7860d412f4fb455651fe. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/rl/loss.py [hunk 1]
```diff
@@ -99,7 +99,7 @@ def compute_loss(
             kl_mask = loss_mask
         else:
             raise ValueError(f"Invalid KL mask type: {loss_config.kl_mask_type}")
-        loss = loss + loss_config.kl_tau * (log_importance_ratio[kl_mask]).sum()
+	loss = loss + loss_config.kl_tau * (log_importance_ratio[kl_mask].sum().detach() * importance_ratio[kl_mask]).sum()
```
[the remark this answers is not in the document]
Comment (bbartoldson) [reply src/prime_rl/trainer/rl/loss.py]: I don't think this is an issue. As shown in the Kimi K2 paper, the log importance ratio is computed at the sequence level, which is why we have the first sum. It is detached because this ratio is inside a square operation (meaning the log ratio will show up in the gradient). Then, you multiply by the importance ratio to simultaneously apply token-level importance sampling and get the gradient w.r.t. the log probs. Finally, you sum everything to get the total loss for the sequence.

image

**Comment by mikasenghaas** [comment]
nice, thx @bbartoldson! im curious about your experimental setup and environment that you use for testing off-policy stability. do you have an env (countdown task?) + config that you could point to that we can use to reproduce the ablation results that you use? might make sense for us to throw this or smth similar into our nightly test suite as well

**Comment by bbartoldson** [comment]
hi @mikasenghaas, happy to provide more details, please lmk if something beyond the following might help.

### compute

4xH100 for training, 4xH100 for inference

### config

I use the PRIME-RL [train](https://github.com/bbartoldson/tba-prime/blob/main/src/zeroband/training/config.py) and [inference](https://github.com/bbartoldson/tba-prime/blob/main/src/zeroband/inference/config.py) defaults, modified as follows.

1. `max_model_len = 4096`
2. `sampling.n = 16`
3. `lr=1e-6`
4. `model="Qwen/Qwen3-4B-Instruct-2507"`

I use the [RatioConfig](https://github.com/bbartoldson/tba-prime/blob/main/src/zeroband/training/loss.py#L306) (CISPO-style clipping with the default `clip_ratio=8`) for the Dr. GRPO baseline. This IS clipping approach is also used for the K2 baseline and TBA.

To start training, I use [this script](https://github.com/bbartoldson/tba-prime/blob/main/multinode_launch.sh), which I give [experiment specific args like this](https://github.com/bbartoldson/tba-prime/blob/main/Countdown_experiments/GRPO_qwen3_async10). I also use a [base training config](https://github.com/bbartoldson/tba-prime/blob/main/configs/training/COUNTDOWN.toml) and a [base inference config](https://github.com/bbartoldson/tba-prime/blob/main/configs/inference/COUNTDOWN.toml). 

### env

i create 10K countdown problems with [this script](https://github.com/bbartoldson/tba-prime/blob/main/scripts/task_generation/reasoning_gym/prepare_dataset.py) and the command

`uv run python scripts/task_generation/reasoning_gym/prepare_dataset.py --size 10000 --task countdown`

then move 1K problems over to a test set. 

### objectives

1. For TBA, you just [add the KL's deviation-from-the-mean-KL to the advantage prior to multiplying by the IS ratio](https://github.com/bbartoldson/tba-prime/blob/main/src/zeroband/training/loss.py#L238), in addition to the [reference resetting](https://github.com/bbartoldson/tba-prime/blob/main/src/zeroband/train.py#L581). 
2. For K2, [you do the same thing with the KL (but no mean subtraction)](https://github.com/bbartoldson/tba-prime/blob/kimi_k2_seq_level/src/zeroband/training/loss.py#L237), and no need to manually reset the reference due to the inference policy's acting as the reference policy.
   - This objective is equivalent to the one in the PR, except it uses CISPO-style clipping (I also [clip token-level KL terms at -10 and +10](https://github.com/bbartoldson/tba-prime/blob/kimi_k2_seq_level/src/zeroband/training/loss.py#L234), but I don't know if this matters at all).
   - The stable K2 results [shown here](https://github.com/PrimeIntellect-ai/prime-rl/issues/1390) use this objective.
3. You can [remove the CISPO clipping and just use IcePop masking](https://github.com/bbartoldson/tba-prime/blob/kimi_k2_seq_level/src/zeroband/training/loss.py#L93) along with the K2 KL approach.
   - This is equivalent to the implementation in the PR.
   - **Please note that this will crash when 10 steps off-policy.** My experiments showed that TBA is fine with IcePop masking's replacing CISPO clipping, but I hadn't tested K2 in this setting until recently. I therefore don't recommend this objective for stability on highly off-policy data. 

#### other variants 

Multiplying the IS ratio by the KL term could be excluded, but this reduces accuracy of K2 (objective 2) by a few percent and TBA (objective 1) by more, while the IcePop-only K2 approach (objective 3) crashes still, [as shown here](https://wandb.ai/bartoldson/Countdown_Kimi_ablate_KL_IS/workspace?nw=nwuserbartoldson).

**Comment by faresobeid** [comment]
Will close this issue as its being fixed with other changes in https://github.com/PrimeIntellect-ai/prime-rl/pull/1404. Thanks for the contribution
